File: go/builtins/funcKeys.go
package builtins

import (
	"github.com/jplorg/jpl/go/v2/jpl"
	"github.com/jplorg/jpl/go/v2/library"
)

var funcKeys = enclose(func(runtime jpl.JPLRuntime, signal jpl.JPLRuntimeSignal, next jpl.JPLPiper, input any, args ...any) ([]any, error) {
	value, err := library.UnwrapValue(input)
	if err != nil {
		return nil, err
	}
	t, err := library.Type(value)
	if err != nil {
		return nil, err
	}

	switch t {
	case jpl.JPLT_ARRAY:
		v := value.([]any)
		result := make([]any, len(v))
		for i := range v {
			result[i] = float64(i)
		}
		return next.Pipe(result)

	case jpl.JPLT_OBJECT:
		v := value.(map[string]any)
		result := make([]any, 0, len(v))
		for k := range v {
			result = append(result, k)
		}
		return next.Pipe(result)

	default:
	}

	return nil, library.ThrowAny(library.NewTypeError("%s (%*<100v) has no keys", string(t), value))
})

File: js/lib/builtins/funcKeys.ts
import type JPLRuntimeSignal from '@/library/runtimeSignal';
import type JPLRuntime from '@/runtime';
import { JPLTypeError } from '../library';

export default function builtin(
  runtime: JPLRuntime,
  _signal: JPLRuntimeSignal,
  next: (output: unknown) => Promise<unknown[]> | unknown[],
  input: unknown,
): Promise<unknown[]> | unknown[] {
  const value = runtime.unwrapValue(input);
  const t = runtime.type(value);

  switch (t) {
    case 'array':
      return next((value as unknown[]).map((_, i) => i));

    case 'object':
      return next(Object.keys(value as Record<string, unknown>));

    default:
  }

  throw new JPLTypeError('%s (%*<100v) has no keys', t, value);
}
